src/wayland/dmabuf/mod.rs

```diff
@@ -0,0 +1,465 @@
+            return;
+        }
+        let info = BufferInfo {
+            planes: ::std::mem::replace(&mut self.pending_planes, Vec::new()),
+            width,
+            height,
+            format,
+            flags,
+        };
+        let mut handler = self.handler.borrow_mut();
+        if let Ok(data) = handler.validate_dmabuf(info) {
+            trace!(self.log, "Creating a new validated immediate dma wl_buffer.");
+            handler.create_buffer(data, buffer_id);
+        } else {
+            trace!(
+                self.log,
+                "Refusing creation of an invalid immediate dma wl_buffer, killing client."
+            );
+            params.as_ref().post_error(
+                ParamError::InvalidWlBuffer as u32,
+                "create_immed resulted in an invalid buffer.".into(),
+            );
+        }
+    }
+}
+
+fn buffer_basic_checks(
+    formats: &[Format],
+    pending_planes: &[Plane],
+    params: &BufferParams,
+    format: u32,
+    width: i32,
+    height: i32,
+) -> bool {
+    // protocol_checks:
+    // This must be a known format
+    let format = match formats.iter().find(|f| f.format == format) {
+        Some(f) => f,
+        None => {
+            params.as_ref().post_error(
+                ParamError::InvalidFormat as u32,
+                format!("Format {:x} is not supported.", format),
+            );
+            return false;;
+        }
+    };
+    // The number of planes set must match what the format expects
+    let max_plane_set = pending_planes.iter().map(|d| d.plane_idx + 1).max().unwrap_or(0);
+    if max_plane_set != format.plane_count || pending_planes.len() < format.plane_count as usize {
+        params.as_ref().post_error(
+            ParamError::Incomplete as u32,
+            format!(
+                "Format {:x} requires {} planes but got {}.",
+                format.format, format.plane_count, max_plane_set
+            ),
+        );
+        return false;
+    }
+    // Width and height must be positivie
+    if width < 1 || height < 1 {
+        params.as_ref().post_error(
+            ParamError::InvalidDimensions as u32,
+            format!("Dimensions ({},{}) are not valid.", width, height),
+        );
+        return false;
+    }
+    // check the size of each plane buffer
+    for plane in pending_planes {
+        // check size for overflow
+        let end = match plane
+            .stride
+            .checked_mul(height as u32)
+            .and_then(|o| o.checked_add(plane.offset))
+        {
+            None => {
+                params.as_ref().post_error(
+                    ParamError::OutOfBounds as u32,
+                    format!("Size overflow for plane {}.", plane.plane_idx),
+                );
+                return false;
+            }
+            Some(e) => e,
+        };
+        if let Ok(size) = ::nix::unistd::lseek(plane.fd, 0, ::nix::unistd::Whence::SeekEnd) {
```

Comment:
Won't we want to reset the Seek ptr? Otherwise this might break, if the user later converts the fd to a `File` or similar.

Reply:
Interesting point. I assumed it was not necessary for drm import later, as weston leaves it as it. I guess it cannot hurt to reset it though.